community/cypher/frontend-3.2/src/main/scala/org/neo4j/cypher/internal/frontend/v3_2/parser/Patterns.scala

```diff
@@ -88,15 +88,22 @@ trait Patterns extends Parser
           RelationshipTypes ~~ MaybeVariableLength ~
           MaybeProperties ~~
         "]"
-      | EMPTY ~ push(None) ~ push(Seq()) ~ push(None) ~ push(None)
+      | EMPTY ~ push(None) ~ push(MaybeLegacyRelTypes()) ~ push(None) ~ push(None)
     )
   }
 
-  private def RelationshipTypes: Rule1[Seq[ast.RelTypeName]] = rule("relationship types") (
-      ":" ~~ oneOrMore(RelTypeName, separator = WS ~ "|" ~~ optional(":") ~ WS)
-    | EMPTY ~ push(Seq())
+  private def RelationshipTypes: Rule1[MaybeLegacyRelTypes] = rule("relationship types") (
+    (":" ~~ RelTypeName ~~ zeroOrMore(WS ~ "|" ~~ LegacyCompatibleRelTypeName)) ~~>> (
+      (first: ast.RelTypeName, more:List[(Boolean,ast.RelTypeName)]) => (pos:InputPosition) => {
+        MaybeLegacyRelTypes(first +: more.map(_._2), more.exists(_._1))
+      })
+    | EMPTY ~ push(MaybeLegacyRelTypes())
   )
 
+  private def LegacyCompatibleRelTypeName: Rule1[(Boolean,ast.RelTypeName)] =
+    ((":" ~ push(true)) | EMPTY ~ push(false)) ~~ RelTypeName ~~>> (
```

Comment:
I guess this is the line that sets the warning condition. So the warning will be emitted when doing

```
MATCH ()-[:FOO | :BAR]->()
```

but not when doing

```
MATCH ()-[:FOO | BAR]->()
```

and the presence of an inlined property predicate is not checked for. It would be nice to have some acceptance tests that confirms this interpretation.